content/05.main-text.md

```diff
@@ -749,10 +767,68 @@ Producing results for these queries relies on internal calls to the PostgreSQL r
+
+We used React's own `create-react-app` command line tool to generate a boilerplate for the app.
+This greatly simplified setting up and maintaining the app's testing and building pipelines, bypassing time-consuming configuration of things like Webpack and linters.
+Some configuration of was necessary to produce non-hashed, consistently named output files like `index.js` that could be easily and reliably referenced by and embedded into the Hetio Jekyll website.
+
+For authoring components, we used React's traditional class syntax.
+At the time of authoring the app, React Hooks were still nascent, thus the simpler and less-verbose functional syntax was not viable.
+
+We created a custom package of React components and utility functions that could be shared across the multiple interactive apps on the website.
+The package consists of interface "components" (building blocks) like buttons, sortable/searchable/paginated tables, etc., as well as utility functions for formatting data, debugging, etc.
+Each of the interactive apps import this package to reduce code repetition, and to enforce a consistent style and behavior across the website.
+
+For managing state in the connectivity search app, we used the Redux library.
+In general, Redux is a well-accepted approach to managing complex state.
+To be more explicit, Redux was chosen over vanilla React or other state management libraries for a few reasons:
+
+1) The state in this app was very "global", meaning most of it was needed by a lot of different parts of the app.
+Redux provides a convenient global "store" of state that is easily accessible to any component, avoiding the "prop-drilling" phenomenon.
+2) The structure of the state is nested and complex.
+Redux's "reducer" approach make it cleaner to modify this type of data.
+3) Redux's approach to immutable state that is updated by actions and pure functions makes the application easier to debug.
+It is easy to get a clear timeline of how and when the state changed, and what changed it. 
+
+To create the graph visualization at the bottom of the app, the D3 library was used.
+D3 was chosen over other many other library choices for flexibility and comprehensiveness of features.
+At the time of development, no other library could be found that satisfied several core requirements:
+
+1) SVG implementation for high-resolution, publication-ready figures.
+2) Force-directed layout for untangling nodes.
+3) Pinnable nodes and other physics customizations.
+4) Customizable node and edge drag/hover/select behavior.
+6) intuitive pan/zoom view that works on desktop and mobile.
+7) node and edge appearances that were completely customizable for alignment, text wrapping, color, outlines, fonts, arrowheads, curved and stacked coincident edges, etc.
+
+<!-- write about challenges of integrating D3 with React? -->
```

Comment:
> write about challenges of integrating D3 with React

Did you develop any novel solutions here? Would this be helpful for other developers to be aware of? Depending on the answers to those questions, a couple sentences might be warranted.

Reply:
Not really, in fact it ended up being quite a pain. In following projects like Adage I sort of figured out a cleaner way that I haven't seen done before. I'll just leave this out.